### PR TITLE
[Do Not Merge] GC: State transfer support for GC.

### PR DESCRIPTION
**Pull request by WenbinZhu**
## Overview

GC: State transfer support for GC.

Why should this be merged: 

Related issue(s) (if applicable): #<number>

## Checklist (Definition of Done):

- [x] There are no TODOs left in the code
- [x] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [x] Change is covered by automated tests
- [x] Public API has Javadoc

### CODE REVIEW

**Comment by annym** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 7
- Added 10
           

Complexity increasing per file
==============================
- runtime/src/main/java/org/corfudb/protocols/wireprotocol/LogRecoveryStateWriteMsg.java  1
- infrastructure/src/main/java/org/corfudb/infrastructure/BatchProcessor.java  1
- runtime/src/main/java/org/corfudb/protocols/wireprotocol/LogRecoveryStateResponse.java  1
         

Complexity decreasing per file
==============================
+ infrastructure/src/main/java/org/corfudb/infrastructure/orchestrator/actions/StateTransfer.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/CorfuDB/CorfuDB/pullRequest?prid=4653404&bid=15584307)

File: infrastructure/src/main/java/org/corfudb/infrastructure/orchestrator/actions/StateTransfer.java [hunk 1]
```diff
@@ -184,29 +176,34 @@ private static void transferChunk(Layout layout,
...
                 chunk.get(0), chunk.get(chunk.size() - 1), (ts2 - ts1));
 
-        List<LogData> entries = new ArrayList<>();
+        List<LogData> logEntries = new ArrayList<>();
         for (long address : chunk) {
-            if (!dataMap.containsKey(address)) {
-                log.error("Missing address {} in batch {}", address, chunk);
-                throw new IllegalStateException("Missing address");
+            LogData logEntry = recoveryState.getLogEntryMap().get(address);
+            if (logEntry != null) {
```
Comment (lixinchengdu) [inline infrastructure/src/main/java/org/corfudb/infrastructure/orchestrator/actions/StateTransfer.java]: comments on why logEntry could be null.

File: infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java [hunk 2]
```diff
@@ -334,37 +356,48 @@ public void multiRead(CorfuPayloadMsg<MultipleReadRequest> msg, ChannelHandlerCo
...
-    public void inspectAddresses(CorfuPayloadMsg<InspectAddressesRequest> msg,
-                                 ChannelHandlerContext ctx, IServerRouter r) {
-        log.trace("inspectAddresses: {}", msg.getPayload().getAddresses());
-        InspectAddressesResponse inspectResponse = new InspectAddressesResponse();
+    @ServerHandler(type = CorfuMsgType.LOG_RECOVERY_STATES_REQUEST)
+    public void readRecoveryStates(CorfuPayloadMsg<MultipleReadRequest> msg,
+                                   ChannelHandlerContext ctx, IServerRouter r) {
+        log.trace("readRecoveryStates: {}", msg.getPayload().getAddresses());
+
+        // Read garbage entries first to avoid losing garbage information since
```
Comment (lixinchengdu) [inline infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java]: In your comment, you said that we should read garbage entry first. However, in the actual implementation, the logData is read first.

File: infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java [hunk 3]
```diff
@@ -290,6 +293,25 @@ private void multiGarbageWrite(CorfuPayloadMsg<MultipleWriteMsg> msg,
                 });
     }
 
+    @ServerHandler(type = CorfuMsgType.LOG_RECOVERY_STATES_WRITE)
+    private void writeRecoveryStates(CorfuPayloadMsg<LogRecoveryStateWriteMsg> msg,
+                                     ChannelHandlerContext ctx, IServerRouter r) {
+        LogRecoveryStateWriteMsg recoveryState = msg.getPayload();
+        log.debug("writeRecoveryStates: Writing {} log entries, {} garbage entries, " +
+                "compaction mark: {}", recoveryState.getLogEntries().size(),
```
Comment (lixinchengdu) [inline infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java]: Why not updating comittedTail

File: infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java [hunk 4]
```diff
@@ -290,6 +293,25 @@ private void multiGarbageWrite(CorfuPayloadMsg<MultipleWriteMsg> msg,
...
+                recoveryState.getLogEntries().size(), recoveryState.getCompactionMark());
+
+        batchWriter.addTask(RECOVERY_STATE_WRITE, msg)
+                .thenRunAsync(() -> {
+                    streamLog.updateCommittedTail(recoveryState.getCompactionMark());
```
[the remark this answers is not in the document]
Comment (WenbinZhu) [reply infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java]: Fixed.

File: infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java [hunk 5]
```diff
@@ -334,37 +356,48 @@ public void multiRead(CorfuPayloadMsg<MultipleReadRequest> msg, ChannelHandlerCo
...
 
-    @ServerHandler(type = CorfuMsgType.INSPECT_ADDRESSES_REQUEST)
-    public void inspectAddresses(CorfuPayloadMsg<InspectAddressesRequest> msg,
-                                 ChannelHandlerContext ctx, IServerRouter r) {
-        log.trace("inspectAddresses: {}", msg.getPayload().getAddresses());
-        InspectAddressesResponse inspectResponse = new InspectAddressesResponse();
+    @ServerHandler(type = CorfuMsgType.LOG_RECOVERY_STATES_REQUEST)
+    public void readRecoveryStates(CorfuPayloadMsg<MultipleReadRequest> msg,
+                                   ChannelHandlerContext ctx, IServerRouter r) {
+        log.trace("readRecoveryStates: {}", msg.getPayload().getAddresses());
```
[the remark this answers is not in the document]
Comment (WenbinZhu) [reply infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java]: This is to make consistent with the normal read API.

File: runtime/src/main/java/org/corfudb/protocols/wireprotocol/LogRecoveryStateResponse.java [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+import java.util.Map;
+
+/**
+ * Created by WenbinZhu on 11/25/19.
+ */
+@Data
+@AllArgsConstructor
+public class LogRecoveryStateResponse implements ICorfuPayload<LogRecoveryStateResponse> {
+
+    Map<Long, LogData> logEntryMap;
```
[the remark this answers is not in the document]
Comment (WenbinZhu) [reply runtime/src/main/java/org/corfudb/protocols/wireprotocol/LogRecoveryStateResponse.java]: Done.

File: runtime/src/main/java/org/corfudb/protocols/wireprotocol/LogRecoveryStateResponse.java [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+    Map<Long, LogData> garbageEntryMap;
```
[the remark this answers is not in the document]
Comment (WenbinZhu) [reply runtime/src/main/java/org/corfudb/protocols/wireprotocol/LogRecoveryStateResponse.java]: Done.

File: runtime/src/main/java/org/corfudb/protocols/wireprotocol/LogRecoveryStateResponse.java [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+
+    long compactionMark;
```
[the remark this answers is not in the document]
Comment (WenbinZhu) [reply runtime/src/main/java/org/corfudb/protocols/wireprotocol/LogRecoveryStateResponse.java]: Done.

File: runtime/src/main/java/org/corfudb/protocols/wireprotocol/LogRecoveryStateResponse.java [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+    }
+
+    public void putGarbageData(Long address, LogData data) {
+        garbageEntryMap.put(address, data);
+    }
+
+    public void merge(LogRecoveryStateResponse other) {
+        logEntryMap.putAll(other.getLogEntryMap());
+        garbageEntryMap.putAll(other.getGarbageEntryMap());
+        compactionMark = Math.max(compactionMark, other.getCompactionMark());
```
[the remark this answers is not in the document]
Comment (WenbinZhu) [reply runtime/src/main/java/org/corfudb/protocols/wireprotocol/LogRecoveryStateResponse.java]: Done.

File: infrastructure/src/main/java/org/corfudb/infrastructure/orchestrator/actions/StateTransfer.java (hunk 1 above)
Comment (WenbinZhu) [reply infrastructure/src/main/java/org/corfudb/infrastructure/orchestrator/actions/StateTransfer.java]: Done.

File: infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java (hunk 2 above)
Comment (WenbinZhu) [reply infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java]: Fixed.

**Pull request closed without merging**